teps/0126-allow-task-sidecars-to-be-specified-in-pipelinerun.md

```diff
@@ -0,0 +1,123 @@
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Extend `PipelineRun` to allow authors to specify sidecars of tasks.
+
+## Motivation
+
+Make tasks more reusable.
+
+Tasks are meant to be shared (e.g. through Git or the ArtifactHub) and referenced from pipeline runs. This TEP seeks to make tasks more often reusable by removing the limitation that sidecars are "hardcoded" and cannot be changed by the pipeline run author.
+
+### Goals
+
+* Extend `PipelineRun` to allow to specify all aspects of sidecars, matching how [sidecars are specified in tasks](https://tekton.dev/docs/pipelines/tasks/#specifying-sidecars).
+
+### Non-Goals
+
+* Pipeline level sidecars, that live for the duration of the whole pipeline run, see https://github.com/tektoncd/pipeline/issues/2973.
+
+### Use Cases
+
+The main use case that triggers this TEP is tasks that run tests. Let's consider the [golang-test](https://artifacthub.io/packages/tekton-task/tekton-catalog-tasks/golang-test) task as an example. It basically runs `go test`. This is fine if the tests executed are true unit tests where all dependencies are mocked. Now consider the case where the tests executed are integration tests, for example running tests against a database (like PostgreSQL or Redis). The current state of Tekton does not provide a way for pipeline run authors to spin up a database or similar for the duration of the task run.
+
+To accomplish the use case right now, I see the following workarounds:
+
+1. run the database outside the pipeline run
+2. ["mis"use a parallel task as a sidecar](https://github.com/tektoncd/pipeline/issues/4235#issuecomment-963204054)
+3. create a new task that has the sidecar baked in
+
+Running the services outside the pipeline run (workaround #1) may require additional work, for example dealing with parallel pipeline runs connecting to the same service. Further, it does not seem right that as a Tekton user I need to setup services outside Tekton to make my Tekton pipeline run work.
+
+Workaround #2 is not seen as adequate because the parallel task has a lifetime longer than the task run, a finally task is needed to terminate it, and it uses the feature "parallel tasks" to mimick another feature ("sidecars") instead of giving access to the sidecar feature where the user needs it.
+
+Creating a task with a sidecar baked in (workaround #3) certainly works, but it prevents reuse: I can't use the `golang-test` task anymore just because there is no sidecar defined. The task I would need to create is exactly the same as `golang-test`, except for the sidecar definition. If there was a way to override sidecars, I could use the task from the ArtifactHub as is and supply the sidecar configuration fitting for the pipeline run.
```

Comment:
I think this is a bit what @jerop was commenting, but if we had a way to "enhance" a Task with steps (pre/post the task) at runtime or authoring time, this one could be achieve while keeping re-usability of Tasks.

At authoring time, as a user, you would "wrap" the `golang-test` with your own set of requirements (expressed as steps before and after the "`golang-test` steps".